Firebase/Storage/FIRStorageDownloadTask.m

```diff
@@ -45,97 +46,102 @@ - (void)enqueue {
-    // Handle file downloads
-    [fetcher setDestinationFileURL:_fileURL];
-    [fetcher setDownloadProgressBlock:^(int64_t bytesWritten, int64_t totalBytesWritten,
-                                        int64_t totalBytesExpectedToWrite) {
-      weakSelf.state = FIRStorageTaskStateProgress;
-      weakSelf.progress.completedUnitCount = totalBytesWritten;
-      weakSelf.progress.totalUnitCount = totalBytesExpectedToWrite;
-      FIRStorageTaskSnapshot *snapshot = weakSelf.snapshot;
-      [weakSelf fireHandlersForStatus:FIRStorageTaskStatusProgress snapshot:snapshot];
-      weakSelf.state = FIRStorageTaskStateRunning;
-    }];
-  } else {
-    // Handle data downloads
-    [fetcher setReceivedProgressBlock:^(int64_t bytesWritten, int64_t totalBytesWritten) {
-      weakSelf.state = FIRStorageTaskStateProgress;
-      weakSelf.progress.completedUnitCount = totalBytesWritten;
-      int64_t totalLength = [[weakSelf.fetcher response] expectedContentLength];
-      weakSelf.progress.totalUnitCount = totalLength;
-      FIRStorageTaskSnapshot *snapshot = weakSelf.snapshot;
-      [weakSelf fireHandlersForStatus:FIRStorageTaskStatusProgress snapshot:snapshot];
-      weakSelf.state = FIRStorageTaskStateRunning;
+    strongSelf.state = FIRStorageTaskStateQueueing;
+    NSMutableURLRequest *request = [strongSelf.baseRequest mutableCopy];
+    request.HTTPMethod = @"GET";
+    request.timeoutInterval = strongSelf.reference.storage.maxDownloadRetryTime;
+    NSURLComponents *components =
+        [NSURLComponents componentsWithURL:request.URL resolvingAgainstBaseURL:NO];
+    [components setQuery:@"alt=media"];
+    request.URL = components.URL;
+
+    GTMSessionFetcher *fetcher;
+    if (resumeData) {
+      fetcher = [GTMSessionFetcher fetcherWithDownloadResumeData:resumeData];
+      fetcher.comment = @"Resuming DownloadTask";
+    } else {
+      fetcher = [strongSelf.fetcherService fetcherWithRequest:request];
+      fetcher.comment = @"Starting DownloadTask";
+    }
+
+    [fetcher setResumeDataBlock:^(NSData *data) {
+      if (data) {
+        FIRStorageDownloadTask *strong = weakSelf;
+        strong->_downloadData = data;
+      }
     }];
-  }
 
-  _fetcher = fetcher;
+    fetcher.maxRetryInterval = strongSelf.reference.storage.maxDownloadRetryTime;
+
+    if (strongSelf->_fileURL) {
+      // Handle file downloads
+      [fetcher setDestinationFileURL:strongSelf->_fileURL];
+      [fetcher setDownloadProgressBlock:^(int64_t bytesWritten, int64_t totalBytesWritten,
+                                          int64_t totalBytesExpectedToWrite) {
+        weakSelf.state = FIRStorageTaskStateProgress;
+        weakSelf.progress.completedUnitCount = totalBytesWritten;
+        weakSelf.progress.totalUnitCount = totalBytesExpectedToWrite;
+        FIRStorageTaskSnapshot *snapshot = weakSelf.snapshot;
+        [weakSelf fireHandlersForStatus:FIRStorageTaskStatusProgress snapshot:snapshot];
+        weakSelf.state = FIRStorageTaskStateRunning;
```

Comment:
This is probably copy-pasted code from before, but imo it makes more sense to use `strongSelf` for these assignments.

Reply:
That would create retain cycles unless I somehow clear the block. Please do correct me if I am mistaken here.

Reply:
Oops, that's correct.